### PR TITLE
Locale support for extent and raster save as dlg

### LINKED ISSUE #41284: raster export (save as...) not working if QGIS locale uses commas as decimal separator
QGIS master Ubuntu 20.04

*) in QGIS properties > general set a locale to language that uses the comma as a decimal separator (pt_PT in my case)

*) add a raster layer

*) right click the raster layer > export > save as...

*) set the export properties and try to set a HxV resolution to a decimal value > QGIS allows to enter the only the comma as a separator

*) click "ok" >

```
Cannot write raster. Error code: data source creation
Cannot create new dataset  /tmp/test.tif:
Attempt to create 0x0 dataset is illegal,sizes must be larger than zero.
```

*) works ok if resolution is set to values with no decimals

I am pretty sure to have seen this (QGIS not letting use the point as a separator in case the QGIS locale uses the comma) in other parts of QGIS but I don't recall where right now.

### PR DESCRIPTION
**Pull request by elpaso**
Fixes #41284

### CODE REVIEW

**Comment by elpaso** [comment]
CC @SrNetoChan , @agiudiceandrea @gioman 

Can you please check this out? Also test nodata and pyramids (no locale support on the latter)

**Comment by elpaso** [comment]
CC @SrNetoChan , @agiudiceandrea @gioman 

Can you please check this out? Also test nodata and pyramids (no locale support on the latter)

**Comment by agiudiceandrea** [comment]
@elpaso, I don't know it is related to this PR, but QGIS (mxe) crashes every time a layer is removed from the map if there is only one layer in the map.

File: src/gui/qgsextentwidget.cpp [hunk 1]
```diff
@@ -416,8 +416,8 @@ void QgsExtentWidget::extentDrawn( const QgsRectangle &extent )
 
 QgsRectangle QgsExtentWidget::outputExtent() const
 {
-  return QgsRectangle( mXMinLineEdit->text().toDouble(), mYMinLineEdit->text().toDouble(),
-                       mXMaxLineEdit->text().toDouble(), mYMaxLineEdit->text().toDouble() );
+  return QgsRectangle( QLocale().toDouble( mXMinLineEdit->text() ), QLocale().toDouble( mYMinLineEdit->text() ),
+                       QLocale().toDouble( mXMaxLineEdit->text() ), QLocale().toDouble( mYMaxLineEdit->text() ) );
```
Comment (agiudiceandrea) [inline src/gui/qgsextentwidget.cpp:420]:
```suggestion
  return QgsRectangle( QgsDoubleValidator::toDouble( mXMinLineEdit->text() ), QgsDoubleValidator::toDouble( mYMinLineEdit->text() ),
                       QgsDoubleValidator::toDouble( mXMaxLineEdit->text() ), QgsDoubleValidator::toDouble( mYMaxLineEdit->text() ) );
```

File: src/gui/qgsrasterlayersaveasdialog.cpp [hunk 2]
```diff
@@ -353,12 +354,12 @@ int QgsRasterLayerSaveAsDialog::nRows() const
 
 double QgsRasterLayerSaveAsDialog::xResolution() const
 {
-  return mXResolutionLineEdit->text().toDouble();
+  return QLocale().toDouble( mXResolutionLineEdit->text() );
```
Comment (agiudiceandrea) [inline src/gui/qgsrasterlayersaveasdialog.cpp:357]:
```suggestion
  return QgsDoubleValidator::toDouble( mXResolutionLineEdit->text() );
```

File: src/gui/qgsrasterlayersaveasdialog.cpp [hunk 3]
```diff
@@ -353,12 +354,12 @@ int QgsRasterLayerSaveAsDialog::nRows() const
...
 }
 
 double QgsRasterLayerSaveAsDialog::yResolution() const
 {
-  return mYResolutionLineEdit->text().toDouble();
+  return QLocale().toDouble( mYResolutionLineEdit->text() );
```
Comment (agiudiceandrea) [inline src/gui/qgsrasterlayersaveasdialog.cpp:362]:
```suggestion
  return QgsDoubleValidator::toDouble( mYResolutionLineEdit->text() );
```

File: src/gui/qgsrasterlayersaveasdialog.cpp [hunk 4]
```diff
@@ -854,7 +855,7 @@ double QgsRasterLayerSaveAsDialog::noDataCellValue( int row, int column ) const
   {
     return std::numeric_limits<double>::quiet_NaN();
   }
-  return lineEdit->text().toDouble();
+  return QLocale().toDouble( lineEdit->text() );
```
Comment (agiudiceandrea) [inline src/gui/qgsrasterlayersaveasdialog.cpp:858]:
```suggestion
  return QgsDoubleValidator::toDouble( lineEdit->text() );
```

**Comment by elpaso** [comment]
> @elpaso, I don't know if it is related to this PR, but QGIS (mxe artifact build from this PR) crashes every time a layer is removed from the map if there is only one layer in the map.

Probably unrelated.

Thank you for the review btw!

**Comment by agiudiceandrea** [comment]
> Probably unrelated.

You are right, The offending PR is https://github.com/qgis/QGIS/pull/41285#issuecomment-771964497. A patch was submitted right now https://github.com/qgis/QGIS/pull/41319.

**Comment by SrNetoChan** [comment]
@elpaso tested your PR on Ubuntu 18.04

Everything works as expected. 

There were also problems in the nodata options. If you used commas to set a nodata interval from 0 to 100,5 , the nodata option would be ignored and the raster was saved without setting those values to NODATA. 

In your PR it seems to be working.

From this experience, I would say we should/could search for more QString::number and text().toDouble() in gui elements as they are probably causing problems somehow, no?

**Comment by elpaso** [comment]
> @elpaso tested your PR on Ubuntu 18.04
> 
> Everything works as expected.
> 

Thanks for checking!

> There were also problems in the nodata options. If you used commas to set a nodata interval from 0 to 100,5 , the nodata option would be ignored and the raster was saved without setting those values to NODATA.
> 
> In your PR it seems to be working.
> 
> From this experience, I would say we should/could search for more QString::number and text().toDouble() in gui elements as they are probably causing problems somehow, no?

Definitely.

**Comment by elpaso** [comment]
unrelated test failure

**Pull request merged**